### ISSUE TITLE
Heatmap (both go and px) not showing text annotations on Hugo site

### ISSUE THREAD

**Issue by rafaelmarino** [plotly/plotly.py#3651]
Hi, my issue appears to be similar to issue #3527, but on that issue they're using different environments (Dash and Jupyter). I figured I'd provide more data over here. My use case is creating plotly.py figures in vscode, then exporting the figures as json to Hugo, and calling them inside markdown files via Hugo shortcodes. This flow is working great so far, except for this issue where heatmap text annotations don't show up in posts. 

**My environment:**
Ubuntu 20.04.4 LTS
Python 3.9.5 (default, Nov 23 2021, 15:27:38) [GCC 9.3.0] on linux
plotly==5.6.0
hugo v0.88.1-5BC54738 linux/amd64 BuildDate=2021-09-04T09:39:19Z VendorInfo=gohugoio
VSCode Version: 1.65.2
Commit: c722ca6c7eed3d7987c0d5c3df5c45f6b15e77d1
Date: 2022-03-10T15:36:26.048Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Linux x64 5.4.0-105-generic

**Test case:** (from https://plotly.com/python/heatmaps/)
```py
import plotly.graph_objects as go

fig = go.Figure(data=go.Heatmap(
                    z=[[1, 20, 30],
                      [20, 1, 60],
                      [30, 60, 1]],
                    text=[['one', 'twenty', 'thirty'],
                          ['twenty', 'one', 'sixty'],
                          ['thirty', 'sixty', 'one']],
                    texttemplate="%{text}",
                    textfont={"size":20}))
```

**Hugo Config:**
```html
 <!-- head tag -->
  {{ if .Params.plotly }}
  <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
  {{ end }}
```
```html
<!-- layouts/shortcodes/plotly.html -->
{{ $json := .Get "json" }} 
{{ $height := .Get "height" | default "200px" }}
<div id="{{$json}}" class="plotly" style="height:{{$height}}; margin-top:-1rem;"></div>
<script>
    Plotly.d3.json({{ $json }}, function (err, fig) {
        Plotly.plot('{{$json}}', fig.data, fig.layout, { responsive: true });
    });
</script>
```
```html
<!-- call from within post.md -->
{{< plotly json="/json/heatmap2.json" height="700px" >}}
```

**My workflow:**
- Create an annotated heatmap fig on a VSCode Python script (using either `graph_objects` or `plotly.express`)
- Create the figure using: `fig.show()`, `fig.write_image("fig.png")` and `fig.write_json(fig.json")`
- Annotations show perfectly for `fig.show()`, `fig.write_image("fig.png")`
- I copy and paste `fig.json` in my `hugo/static/` folder
- The heatmap shows, but without annotations, see attachment.
![image](https://user-images.githubusercontent.com/9305283/161104493-7e2e1a17-fd1e-471d-ae6f-0ce3c2b3881a.png )

Thanks for reading. Happy to help with any tests.

**Comment by derek-pyne** [comment]
Think the issue is that `https://cdn.plot.ly/plotly-latest.min.js` is not actually pointing to the latest Plotly version. Open it in a browser and you'll see its `1.58.5`.

If you replace it with `https://cdn.plot.ly/plotly-2.11.1.min.js` you'll get the latest Plotly (which you need since these annotations are a newish feature) and replace `Plotly.plot` with `Plotly.newPlot` you should be good to go

**Comment by rafaelmarino** [comment]
Thanks @derek-pyne, that's an interesting find. Unfortunately that didn't fix it for me. After switching to `Plotly.newPlot` and `https://cdn.plot.ly/plotly-2.11.1.min.js` plotly plots stop rendering altogether, instead showing a blank, placeholding space where the figures were.

> Note: Switching from `Plotly.plot` to `Plotly.newPlot` while keeping `https://cdn.plot.ly/plotly-latest.min.js` does render the plots same as before, but the issue of missing heatmap annotations persists.

**Comment by derek-pyne** [comment]
Any errors in your browsers console? (In Chrome, right-click `Inspect` then go to the `Console` tab)

**Comment by derek-pyne** [comment]
Also, you need to use the `2.11.1` link, not the `latest link`.

`Plotly.plot` is when using `latest` (which is actually 1.58.5)
`Plotly.newPlot` is when using `2.11.1`

**Comment by rafaelmarino** [comment]
Yeah, I was just noting that `Plotly.newPlot` seemingly works with `latest` (`1.58.5`), since plots do render.

You were right about errors in the console log, see below (this is with `Plotly.newPlot` and `2.11.1`):

I get one of this for each one of my plots (adapted with their corresponding locations):

```
(index):206 
Uncaught TypeError: Cannot read properties of undefined (reading 'json')
 at (index):206:15 (anonymous) @ (index):206
```
Clicking on `(index):206:15` leads me to an active cursor placed as shown below:

```
<script>
    Plotly.d3.< cursor placed here >json("/json/f12021/heatmap2.json", function (err, fig) {
        Plotly.newPlot('\/json\/f12021\/heatmap2.json', fig.data, fig.layout, { responsive: true });
    });
</script></p>
```

**Comment by derek-pyne** [comment]
Seems that `Plotly.d3.json` has changed in `2.11.1`... maybe its a different method now?

But I think the core of it is that you've got to get to the newer versions of Plotly (2.11.1) to get the text annotations on heatmaps to work. And since its a big jump in versions (from 1.58 to 2.11), there have been some changes to the library that you have to deal with along the way.

I have a very similar setup to you, except we do our JSON conversion in Python that is working find... so you should be able to find a way :)

**Comment by rafaelmarino** [comment]
Firefox is a bit more clear with the console logs: `Uncaught TypeError: Plotly.d3 is undefined`. 

Basically the issue is a rehash of this 2021 issue (which got zero traction): [Got “Uncaught TypeError: Plotly.d3 is undefined”](https://community.plotly.com/t/got-uncaught-typeerror-plotly-d3-is-undefined/56020): `Plotly.d3` is defined in `latest`&mdash;which is not really the latest, hence lack of annotation support&mdash;yet it's not defined in the actual latest (`2.11.1`). 

Maybe there's a reason for this? Perhaps `latest` is actually `latest-we-want-you-using` or something, and it's deliberately kept pointing to some prior version to not break things. 

I played a bit with versions, and it seems `Plotly.d3` was supported until `1.9.0` (even though plots don't render well), and then it's gone. Anyway, I'm giving up on heatmap annotations for the time being :)

Thanks Derek

**Comment by nicolaskruchten** [comment]
we pinned `-latest` to the last release in the 1.x series because we made breaking changes in the 2.x series (e.g. removing D3 exports!) and didn't want to break anyone's code. There is no more continually-updated `latest` URL: we recommend folks knowingly choose a specific versioned bundle to load. This was all explained in the [Plotly.js 2.0 announcement](https://community.plotly.com/t/announcing-plotly-js-2-0/53675) :)

**Comment by rafaelmarino** [comment]
Thanks Nicolas, that makes sense. I read through some of that, followed a few links, and I'm happy to report I've managed to get it working :)

For those who stumble upon this thread in the future, here's what I did:

1) Update `head.html` with:

```
  <script src="https://d3js.org/d3.v7.min.js"></script>
  <script src="https://cdn.plot.ly/plotly-2.11.1.min.js"></script>
```

2) Update `plotly.html` with:

```
d3.json({{ $json }}).then(function (fig) {
        Plotly.newPlot('{{$json}}', fig.data, fig.layout, { responsive: true });
});
```

**Comment by shouldsee** [comment]
> Think the issue is that `https://cdn.plot.ly/plotly-latest.min.js` is not actually pointing to the latest Plotly version. Open it in a browser and you'll see its `1.58.5`.

This is weird. Is this not fixed yet?

**Comment by nicolaskruchten** [comment]
This is indeed awkward but we chose this path in order not to impose breaking changes on folks relying on the "latest" URL so we will not be changing the version of Plotly.js that it points to any more. We apologize for the confusion.

**Comment by shouldsee** [comment]
> This is indeed awkward but we chose this path in order not to impose breaking changes on folks relying on the "latest" URL so we will not be changing the version of Plotly.js that it points to any more. We apologize for the confusion.

Gotcha. Thanks for explaining!
